### DISCUSSION TITLE
Project Sync Window ... what does duration do ?

### DISCUSSION THREAD

**Discussion by adrianmiron** [argoproj/argo-cd#16005] [Q&A]
Hi,

I am trying to understand what is the scope of the `duration` field in the sync window feature :
https://argo-cd.readthedocs.io/en/stable/user-guide/sync_windows/

I get the cron part, which defines the interval, but all i got about duration is that it extends the interval outside the cron interval ? Is that it ?

Thanks

**Comment by pasha-codefresh** [comment]
This is a code that counting it 


```
	// If SyncWindow.Active() is called outside of a UTC locale, it should be
	// first converted to UTC before search
	currentTime = currentTime.UTC()

	specParser := cron.NewParser(cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow)
	schedule, _ := specParser.Parse(w.Schedule)
	duration, _ := time.ParseDuration(w.Duration)

	// Offset the nextWindow time to consider the timeZone of the sync window
	timeZoneOffsetDuration := w.scheduleOffsetByTimeZone()
	nextWindow := schedule.Next(currentTime.Add(timeZoneOffsetDuration - duration))

	return nextWindow.Before(currentTime.Add(timeZoneOffsetDuration))

```

And short explanation

**nextWindow := schedule.Next(currentTime.Add(timeZoneOffsetDuration - duration)):** 
This line calculates the next scheduled time (next window) based on the provided schedule. It adjusts the calculation by subtracting the duration and applying the timeZoneOffsetDuration. Essentially, it determines when the next scheduled event should occur, taking into account the time zone offset and the duration of the event.

**return nextWindow.Before(currentTime.Add(timeZoneOffsetDuration)):** 
Finally, this line checks if the calculated nextWindow is before the current time adjusted for the time zone offset. If the nextWindow is earlier than the current time plus the time zone offset, it implies that the scheduled event is due to happen soon, and the function returns true. Otherwise, it returns false.

**Reply by adrianmiron** [reply]
Thanks for the details, let me see if i understood.

My set TZ is UTC, so no offset required.
My schedule windows schedule is: 0-6 * * 1-5:1h

Between 0 A.M. and 6 A.M i assume the window is active and the `duration` field is telling the controller how much time is left before the next window ,  so at 23 P.M. it would tell it it's 1h before  ?

Why is that relevant ? How often is the evaluation done ? How would it help me and/or the controller to set 1h vs 10m in the duration parameter ?

Thanks

**Reply by pasha-codefresh** [reply]
1.  How often is the evaluation done, this evaluation happens each time when you trying to sync 
2. Probably i am wrong , but you cant define such cron expressions as you provided, you have to define like this cron.NewParser(cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow)

<img width="1185" alt="Снимок экрана 2023-10-19 в 17 22 33" src="https://github.com/argoproj/argo-cd/assets/39732895/8f7f8601-976a-42f6-b61c-b86e1e43007d">


In this screenshot you can see example that you deny to deploy every Sunday and Saturday in 21:00 for 30 minutes, so if you want to deny for 3 hours till 24 , you have to define duration in 3 hours. I hope it is better explanation

**Reply by adrianmiron** [reply]
Well you can define intervals, just shift select multiple hours. The example is my actual configuration from the CRD, test it out yourself.

But i understand your point, i define the start time and duration....i will test that out myself.
